lib/dashing/app.rb

```diff
@@ -70,13 +70,23 @@ def authenticated?(token)
   redirect "/" + dashboard
 end
 
+
 get '/events', provides: 'text/event-stream' do
   protected!
   response.headers['X-Accel-Buffering'] = 'no' # Disable buffering for nginx
-  stream :keep_open do |out|
-    settings.connections << out
+  stream do |out|
     out << latest_events
-    out.callback { settings.connections.delete(out) }
+    settings.connections << connection = {out: out, mutex: Mutex.new, terminated: false}
+    terminated = false
+
+    loop do
+      connection[:mutex].synchronize do
+        terminated = true if connection[:terminated]
+      end
+      break if terminated
+    end
+
```

Comment:
Each request gets a thread in the thread pool until Puma hits the max and multiplexes the threads. The loop is basically just "keep the SSE request open until marked as terminated".

The connection is marked as terminated in `send_event` when there is an exception caught when trying to send to a dead connection. `send_even` is triggered from Rufus -- which is multithreaded in Puma.